### PR TITLE
Allow super in methods with dynamic names

### LINKED ISSUE #1392: calling `super` in methods defined on namespaced classes
`C::m = -> super` compiles correctly into

    C.prototype.m = function() {
      return C.__super__.m.apply(this, arguments);
    };

`namespaced.C::m = -> super"` compiles incorrectly into

    namespaced.C.prototype.m = function() {
      return m.__super__.constructor.apply(this, arguments);
    };

Proper compilation would have been

    var _ref;
    _ref = namespaced.C;
    _ref.prototype.m = function() {
      return _ref.__super__.constructor.apply(this, arguments);
    };

The error originates from [the `METHOD_DEF` regex](https://github.com/jashkenas/coffee-script/blob/085874d5f35835bc8e61/src/nodes.coffee#L1807) being used by [`Assign::compileNode`](https://github.com/jashkenas/coffee-script/blob/085874d5f35835bc8e61/src/nodes.coffee#L935) matching only identifiers. [`Call::superReference`](https://github.com/jashkenas/coffee-script/blob/085874d5f35835bc8e61/src/nodes.coffee#L466) is then compiling incorrectly.

### PR DESCRIPTION
**Pull request by lydell**
As discussed in https://github.com/jashkenas/coffeescript/issues/3039#issuecomment-68916918.
This is the first step to implement dynamic object literal keys (see #3597).

This also fixes #1392.

In short, `super` is now allowed:

    # in class definitions:
    class A
      instanceMethod: -> super
      @staticMethod: -> super
      @staticMethod2 = -> super

    # in assignment where the next to last property is 'prototype':
    A::m = -> super
    A['prototype'].m = -> super
    a.b()[5]::m = -> super
    A::[x()] = -> super
    class B
      @::m = -> super

### CODE REVIEW

**Comment by lydell** [comment]
Since

```coffee
class A
  @a = -> super
```

is allowed, and the static method compiles to

```js
  A.a = function() {
    return A.__super__.constructor.a.apply(this, arguments);
  };
```

shouldn’t

```coffee
A.a = -> super
```

be allowed too, and compile to the same thing? This would allow:

```coffee
# class style
class A extends B
  @a: -> super
  a: -> super

# classic style
A = ->
A extends B
A.a = -> super # currently an error
A::a = -> super

File: src/nodes.coffee [hunk 1]
```diff
@@ -2296,3 +2303,10 @@ unfoldSoak = (o, parent, name) ->
   parent[name] = ifn.body
   ifn.body = new Value parent
   ifn
+
+# Check if `.prop` or `["prop"]` is equal to `property`.
```
Comment (lydell) [inline src/nodes.coffee]: Oops, this comment is wrong.

**Comment by lydell** [comment]
I’ve fixed the bad comment now.

**Comment by lydell** [comment]
My proposal above might not be a good idea, because then

```coffee
A['prototype'].a = -> super
```

and

```coffee
prototype = 'prototype'
A[prototype].a = -> super
```

would compile to different things, which might be confusing. An error message in the latter case might be preferable.

**Comment by lydell** [comment]
Fixed merge conflict.

**Comment by lydell** [comment]
I’ve removed the code to allow `A['prototype'].m=->super`; see #3796. This effectively turns this PR into a code cleanup plus implementation of agreed-upon features:

- https://github.com/jashkenas/coffeescript/issues/3039#issuecomment-68916918
- #3597 

In other words, nothing controversial. This should make it easier to merge, I think.

**Comment by jashkenas** [comment]
Rebase for a clean merge, and we'll merge?

**Comment by lydell** [comment]
Rebased.

**Comment by jashkenas** [comment]
Merged.

**Comment by jashkenas** [comment]
Bravo on killing off that nasty METHOD_DEF, by the way.

**Pull request merged**